File: sourcestats/views/api/maps.py
# Source Server Stats
# File: sourcestats/views/api/maps.py
# Desc: map API views

from flask import jsonify, request
from elasticquery import Filter

from ...app import app
from ...util import get_source_apps
from ...util.elastic import get_es_history, get_es_terms, get_request_filters


@app.route('/api/v1/maps')
def get_maps():
    '''List current maps and the number of servers playing on them.'''
    maps, total = get_es_terms(
        'map',
        filters=get_request_filters(),
        size=request.args.get('size')
    )

    return jsonify(maps=maps, total=total)


@app.route('/api/v1/map/<name>')
def get_map(name):
    '''Gets details about a single map.'''
    filters = get_request_filters()
    filters.append(Filter.term('map', name))

    # Get games/server counts currently playing this map
    games, _ = get_es_terms(
        'game_id',
        filters=filters
    )

    # Attach game names
    apps = get_source_apps()
    games = [
        ((game[0], apps.get(game[0], 'Unknown')), game[1])
        for game in games
    ]

    return jsonify(name=name, games=games)


@app.route('/api/v1/map/<name>/history/players')
def get_map_player_history(name):
    '''Returns a date histogram of players on this map.'''
    filters = get_request_filters()
    filters.append(Filter.term('map', name))

    date_histogram = get_es_history('player_count', filters)

    return jsonify(players=date_histogram)

File: sourcestats/webpack/actions/maps.js
// Source Server Stats
// File: sourcestats/webpack/actions/maps.js
// Desc: actions for the maps list

import 'whatwg-fetch';
import URI from 'URIjs';

export const FETCH_MAPS = 'FETCH_MAPS';
export const FETCH_MAP = 'FETCH_MAP';
export const FETCH_MAP_PLAYER_HISTORY = 'FETCH_MAP_PLAYER_HISTORY';


/** Fetches maps with a game filter. */
export function fetchMaps(gameId = null) {
    let url = new URI('/api/v1/maps');
    let params = {
        size: 0
    };

    if (gameId)
        params.game_id = gameId;

    return dispatch => {
        fetch(url.query(params))
        .then(response => response.json())
        .then(response => {
            dispatch({
                type: FETCH_MAPS,
                maps: response.maps,
                total: response.total,
                gameId: gameId
            });
        });
    };
}


/** Fetches details for a single map. */
export function fetchMap(name) {
    let url = new URI(`/api/v1/map/${name}`);

    return dispatch => {
        fetch(url)
        .then(response => response.json())
        .then(response => {
            dispatch({
                type: FETCH_MAP,
                map: response
            });
        });
    };
}


/** Fetches a date histogram of player counts for a single map. */
export function fetchMapPlayerHistory(name, filters) {
    let url = new URI(`/api/v1/map/${name}/history/players`);

    return dispatch => {
        fetch(url.query(filters))
        .then(response => response.json())
        .then(response => {
            dispatch({
                type: FETCH_MAP_PLAYER_HISTORY,
                players: response.players
            });
        });
    };
}
